### PR TITLE
WIP: [V2] CFE-971: create and mirror a graph image for ocp releases

### PR DESCRIPTION
**Pull request by sherine-k**
# Description

When the user specifies `graph: true` in the imageSetConfig (`mirror.platform.graph`), oc-mirror generates a graph-image (following directives from [this doc](https://docs.openshift.com/container-platform/4.13/updating/updating-restricted-network-cluster/restricted-network-update-osus.html#update-service-graph-data_updating-restricted-network-cluster-osus) and pushes it to the specified disk or mirror.

In order to have retro-compatibility in v2, and in order to reduce the module dependencies of oc-mirror, the implementation suggested in this PR is based on [containers/buildah](https://github.com/containers/buildah), instead of the existing implementation with [go-containerregistry](https://github.com/google/go-containerregistry) and [crane](https://github.com/google/go-containerregistry/tree/main/cmd/crane)

Fixes [CFE-971](https://issues.redhat.com/browse/CFE-971)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

With the following imageSetConfig:
```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
mirror:
 platform:
   channels:
   - name: stable-4.13
     minVersion: 4.13.10
     maxVersion: 4.13.10
   graph: true
```

We can run oc-mirror for MirrorToDisk first: This will create a `graph-image` and save it to local storage
```
$ ./bin/oc-mirror --v2 -c cfe-969.yml -p 6000 file:///home/skhoury/release
... 
2023/10/25 10:56:51  [INFO]   : mode mirrorToDisk 
2023/10/25 10:56:51  [INFO]   : local storage registry will log to /home/skhoury/go/src/github.com/openshift/oc-mirror/logs/registry.log
2023/10/25 10:56:51  [INFO]   : starting local storage on :6000
d
...
2023/10/25 10:56:52  [INFO]   : creating graph data image
2023/10/25 10:56:52  [INFO]   : graph image localhost:6000/graph-image@sha256:41f0b09707f893900679adc712e78f7f73d7d8b245bf68c921726b43743373e0 created and pushed to cache.
2023/10/25 10:56:52  [INFO]   : total release images to copy 184
...
```

Next, we can perform a DiskToMirror, which would push the graph-image, along with release images, to the mirror
```
$ ./bin/oc-mirror --v2 -c cfe-969.yml -p 6000 --from file:///home/skhoury/release docker://localhost:5000/cfe971 --dest-tls-verify=false
```

## Expected Outcome

After mirroring is complete, we need to check the graph-image:
* The contents of /var/lib/cincinnati-graph-data need to belong to `root`.
* The CMD of the image needs to be compliant with the documentation 

This can be achieved like this for example:

```bash
╭─╼skhoury@fedora ~/.../openshift/oc-mirror 
╰─ (CFE-971)-$ podman pull localhost:5000/cfe971/graph-image:latest
Trying to pull localhost:5000/cfe971/graph-image:latest...
Getting image source signatures
Copying blob 3c5f5168e8c4 done   | 
Copying blob cc7c08d56aad skipped: already exists  
Copying config fd8881a489 done   | 
Writing manifest to image destination
fd8881a489058a887b1f82311762f1289f1434b67f83463ec78f596cbb9dd966
╭─╼skhoury@fedora ~/.../openshift/oc-mirror 
╰─ (CFE-971)-$ podman run -it "localhost:5000/cfe971/graph-image:latest" /bin/bash
[root@96b2ab6ddc60 /]# ls -l /var/lib/cincinnati-graph-data/
total 16
-rw-rw-r--. 1 root root 11357 Oct 25 07:26 LICENSE
drwxr-xr-x. 1 root root 31668 Oct 25 07:48 blocked-edges
drwxr-xr-x. 1 root root  1410 Oct 25 07:48 channels
drwxr-xr-x. 1 root root    26 Oct 25 07:48 raw
-rw-rw-r--. 1 root root     6 Oct 25 07:26 version
[root@96b2ab6ddc60 /]# exit
exit
╭─╼skhoury@fedora ~/.../openshift/oc-mirror 
╰─ (CFE-971)-$ podman inspect "localhost:5000/cfe971/graph-image:latest" | jq .[0].Config.Cmd
[
  "/bin/bash",
  "-c",
  "exec cp -rp /var/lib/cincinnati-graph-data//* /var/lib/cincinnati/graph-data"
]
```

### CODE REVIEW

File: cmd/oc-mirror/main.go [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
+	"github.com/containers/storage/pkg/unshare"
 	"github.com/openshift/oc-mirror/pkg/cli/mirror"
 	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
 )
 
 func main() {
+	// oc-mirror runs in rootless mode. It is therefore necessary to
+	// ensure that oc-mirror is re-executed in a user namespace where
+	// it has root privileges.
+	if buildah.InitReexec() {
```
Comment (lmzuccarelli) [inline cmd/oc-mirror/main.go:14]: @sherine-k - should we not gate this ? i.e move it to the executor.go , and only set this if we have graph: true set

File: go.mod [hunk 2]
```diff
@@ -5,10 +5,10 @@ go 1.20
 require (
```
Comment (lmzuccarelli) [inline go.mod:5]: again the joys of go.mod -  over 1000 files changed - roll on go workspaces

**Comment by lmzuccarelli** [comment]
I see the e2e test failing here 

```
reading subuid mappings for user \"1003670000\" and subgid mappings for group \"1003670000\": no subuid ranges found for user \"1003670000\" in /etc/subuid"
```

File: go.mod (hunk 2 above)
Comment (sherine-k) [reply go.mod:5]: I know :/
I separated most of those in a specific commit to facilitate readability of the PR.
This makes it a higher risk change. cc @zhouying7780 @kasturinarra

**Comment by sherine-k** [comment]
> I see the e2e test failing here
> 
> ```
> reading subuid mappings for user \"1003670000\" and subgid mappings for group \"1003670000\": no subuid ranges found for user \"1003670000\" in /etc/subuid"
> ```

E2E is failing because these tests would be running in a container on a k8s cluster.
Something like the following would allow it to run, but that would require changes in the ci-configs...
```
securityContext:
       capabilities:
         add:
           - CAP_SETGID
           - CAP_SETUID
```

I'll privilege first a way to use unshare only in v2, so that e2e aren't affected for the moment.

File: v2/pkg/release/interfaces.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package release
```
Comment (aguidirh) [inline v2/pkg/release/interfaces.go:1]: Thanks @sherine-k for separating consts and interfaces

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -42,6 +43,17 @@ Please submit bug reports as GitHub Issues. When submitting bug reports, please
...
 
+#### Debugging oc-mirror
+Since end of November 2023, and PR [#721](https://github.com/openshift/oc-mirror/pull/721), oc-mirror's v2 implementation is relying on [Buildah](https://github.com/containers/buildah) in order to
+build Cincinnati graph images. 
+
+Since oc-mirror needs to run rootless, it is now re-executing itself in a separate user namespace where buildah can have root privileges.
+
+This modifies the debugging process: In order to debug you should launch dlv debugger for oc-mirror in a user namespace: 
+```
+podman unshare dlv debug cmd/oc-mirror/main.go -- --v2 -c cfe-969.yml --from file:///home/skhoury/release docker://localhost:5000/cfe971/ --dest-tls-verify=false
```
Comment (aguidirh) [inline CONTRIBUTING.md:54]:
```suggestion
podman unshare dlv debug --headless --listen=:43987 cmd/oc-mirror/main.go -- --v2 -c cfe-969.yml --from file:///home/skhoury/release docker://localhost:5000/cfe971/ --dest-tls-verify=false
```

Once the command above is triggered it is possible to use delve to debug (either using dlv directly or attaching to it with a client).

If you use VSCode, it is possible to attach it to the dlv process running in the background, in order to do that it is necessary to add the following code to the configurations[] inside of the launch.json

```
{
    "name": "Attach Package",
    "type": "go",
    "debugAdapter": "dlv-dap",
    "request": "attach",
    "mode": "remote",
    "host": "localhost",
    "port": 43987,
},
{
    "name": "Attach Tests",
    "type": "go",
    "debugAdapter": "dlv-dap",
    "request": "attach",
    "mode": "remote",
    "host": "localhost",
    "port": 43987,
}

```
```suggestion
podman unshare dlv debug cmd/oc-mirror/main.go -- --v2 -c cfe-969.yml --from file:///home/skhoury/release docker://localhost:5000/cfe971/ --dest-tls-verify=false
```

Comment (aguidirh) [inline CONTRIBUTING.md:54]:
```suggestion
podman unshare dlv debug --headless --listen=:43987 cmd/oc-mirror/main.go -- --v2 -c cfe-969.yml --from file:///home/skhoury/release docker://localhost:5000/cfe971/ --dest-tls-verify=false


Once the command above is triggered it is possible to use delve to debug (either using dlv directly or attaching to it with a client).

If you use VSCode, it is possible to attach it to the dlv process running in the background, in order to do that it is necessary to add the following code to the configurations[] inside of the launch.json

```
{
    "name": "Attach Package",
    "type": "go",
    "debugAdapter": "dlv-dap",
    "request": "attach",
    "mode": "remote",
    "host": "localhost",
    "port": 43987,
},
{
    "name": "Attach Tests",
    "type": "go",
    "debugAdapter": "dlv-dap",
    "request": "attach",
    "mode": "remote",
    "host": "localhost",
    "port": 43987,
}

```

```

Comment (sherine-k) [reply CONTRIBUTING.md:54]: Great ! thanks for finding this Alex!

**Comment by sherine-k** [comment]
Team decision is to use a different implementation (with `go-containerregistry`): 
This `containers/buildah` implementation has heavy consequences on debugging and running e2e tests in CI

**Pull request closed without merging**